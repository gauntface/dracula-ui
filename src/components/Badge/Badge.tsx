import React, { HTMLAttributes } from 'react'
import { colors as backgroundColors } from '../Color/Box'
import cx from 'classnames'
import { Text, TextColors } from '../Typography/Text'

const variants = {
  normal: 'drac-badge',
  subtle: 'drac-badge-subtle',
  outline: 'drac-badge-outline'
}

export interface BadgeProps extends HTMLAttributes<HTMLSpanElement> {
  themeColor?: keyof typeof backgroundColors
  variant?: keyof typeof variants
  disabled?: boolean
}

export const Badge: React.FC<BadgeProps> = props => {
  const isOutline = props.variant === 'outline'
  const isSubtle = props.variant === 'subtle'
  const overrideTextColor = isOutline || isSubtle

  const textColorClass = overrideTextColor
    ? TextColors[props.themeColor ?? 'green']
    : undefined

  let backgroundClass = backgroundColors[props.themeColor ?? 'green']
  if (isSubtle) {
    backgroundClass = `${backgroundClass}-transparent`
  }

  const classes = cx(
    'drac-badge',
    backgroundClass,
    variants[props.variant ?? 'normal'],
    textColorClass
  )

  return (
    <span className={classes} {...props}>
      <Text color={overrideTextColor ? props.themeColor : undefined}>
        {props.children}
      </Text>
    </span>
  )
}
