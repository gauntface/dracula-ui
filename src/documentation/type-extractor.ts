import * as ts from 'typescript'

export function extractWithComment(
  fileNames: string[],
  options: ts.CompilerOptions
): void {
  const program = ts.createProgram(fileNames, options)
  const checker: ts.TypeChecker = program.getTypeChecker()

  for (const sourceFile of program.getSourceFiles()) {
    if (!sourceFile.isDeclarationFile) {
      ts.forEachChild(sourceFile, visit)
    }
  }

  function visit(node: ts.Node) {
    const count = node.getChildCount()

    if (count > 0) {
      ts.forEachChild(node, visit)
    }

    if (ts.isInterfaceDeclaration(node) && node.name) {
      const symbol = checker.getSymbolAtLocation(node.name)
      if (symbol) {
        return serializeSymbol(symbol)
      }
    }
  }

  function serializeSymbol(symbol: ts.Symbol) {
    const s = {
      name: symbol.getName(),
      comment: ts.displayPartsToString(symbol.getDocumentationComment(checker)),
      type: checker.typeToString(
        checker.getTypeOfSymbolAtLocation(symbol, symbol.valueDeclaration!)
      )
    }
    console.log(s)
    return s
  }
}

console.log(extractWithComment(['src/components/Avatar/Avatar.tsx'], {}))
