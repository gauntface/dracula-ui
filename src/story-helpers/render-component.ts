import testRenderer from 'react-test-renderer'

export type ComponentExample = {
  code: ReturnType<testRenderer.ReactTestRenderer['toJSON']>
  docs: string
  title: string
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: React.ReactElement,
  docs: string = ''
): ComponentExample {
  const rendered = testRenderer.create(comp)
  const code = rendered.toJSON()
  return {
    title,
    code,
    docs
  }
}
