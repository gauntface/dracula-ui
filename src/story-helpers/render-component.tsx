import React from 'react'
import Enzyme, { mount,  shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import prettier from 'prettier'
import toJson from "enzyme-to-json";

Enzyme.configure({ adapter: new Adapter() })

export type ComponentExample = {
  html: string
  docs: string
  title: string
  react?: any
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: React.ReactElement,
  docs: string = ''
): ComponentExample {

  
  const html = mount(comp).html()
  const react = toJson(shallow(<div>{comp}</div>))

  const prettyHTML = prettier.format(html, {
    parser: 'html'
  })

  return {
    title,
    html: prettyHTML,
    react,
    docs
  }
}
