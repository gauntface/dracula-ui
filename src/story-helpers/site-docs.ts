import {
  renderSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import React from 'react'

// import * as serializer from 'jest-snapshot-serializer-raw'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  describe(`Site: ${componentClass.displayName ?? componentClass.name}`, () => {
    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp(),
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
    })

    const variations = documentation.variations()

    variations.forEach(variation => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp(),
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()
      })
    })
  })
}
