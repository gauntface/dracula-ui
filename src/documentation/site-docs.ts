import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '../story-helpers/render-component'
import React from 'react'
import fs from 'fs-extra'
import { pretty } from '../story-helpers/pretty'
import { getDocGen } from '../documentation/docgen/doc-generator'
import { toDSP } from '../documentation/dsp/component-generator'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name

  describe(`Site: ${name}`, () => {
    let examples: Record<string, ComponentExample> = {}

    afterAll(async () => {
      const docGen = getDocGen(name)
      const dsp = toDSP(name, examples, docGen)
      const dspPath = `${process.cwd()}/dsp/data/components/${name}.json`
      await fs.writeFile(dspPath, pretty(JSON.stringify(dsp), 'json'))
    })

    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
