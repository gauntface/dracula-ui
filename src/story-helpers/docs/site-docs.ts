import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import React from 'react'
import fs from 'fs'
import { pretty } from '@/story-helpers/pretty'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name

  describe(`Site: ${name}`, () => {
    let examples: Record<string, ComponentExample> = {}

    afterAll(() => {
      const path = `${process.cwd()}/examples/${name}.json`
      fs.writeFileSync(path, pretty(JSON.stringify(examples), 'json'))
    })

    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
