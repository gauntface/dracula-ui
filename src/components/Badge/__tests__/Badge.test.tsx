import { docs } from '@/documentation/site-docs'
import React from 'react'
import { snapshot } from '../../../documentation/render-component'
import { Badge } from '../Badge'

docs(Badge, {
  basic: () =>
    snapshot(
      'Usage',
      () => <Badge>Badge</Badge>,
      `You can create a new Badge with any text content you'd like to highlight.`
    ),
  variations() {
    const colors = snapshot(
      'Colors',
      () => <Badge themeColor="animated">Fun Badge</Badge>,
      'Badges can be customized to use any of the Dracula UI theme colors, including all solid, gradient, and animated colors.'
    )

    const variants = snapshot(
      'Variants',
      () => (
        <div>
          <Badge themeColor="purpleCyan" variant="outline">
            Outline
          </Badge>

          <Badge themeColor="purple" variant="subtle">
            Subtle
          </Badge>
        </div>
      ),
      "Use `outline` or `subtle` when you don't want your Badges to stand out, but still want to keep them playful."
    )

    return [colors, variants]
  }
})
