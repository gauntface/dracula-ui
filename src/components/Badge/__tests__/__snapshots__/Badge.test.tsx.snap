// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site: Badge Basic Usage 1`] = `
Object {
  "docs": "You can create a new Badge with any text content you'd like to highlight.",
  "html": "<span class=\\"drac-badge drac-bg-black\\"
  ><span class=\\"drac-text drac-text-white\\">Badge</span></span
>
",
  "react": "<Badge>Badge</Badge>;
",
  "title": "Usage",
}
`;

exports[`Site: Badge Colors 1`] = `
Object {
  "docs": "Badges can be customized to use any of the Dracula UI theme colors, including all solid, gradient, and animated colors.",
  "html": "<span class=\\"drac-badge drac-bg-animated\\" themecolor=\\"animated\\"
  ><span class=\\"drac-text drac-text-white\\">Fun Badge</span></span
>
",
  "react": "<Badge themeColor=\\"animated\\">Fun Badge</Badge>;
",
  "title": "Colors",
}
`;

exports[`Site: Badge Variants 1`] = `
Object {
  "docs": "Use \`outline\` or \`subtle\` when you don't want your Badges to stand out, but still want to keep them playful.",
  "html": "<div>
  <span
    class=\\"drac-badge drac-bg-purple-cyan drac-badge-outline drac-text-purple-cyan\\"
    themecolor=\\"purpleCyan\\"
    variant=\\"outline\\"
    ><span color=\\"purpleCyan\\" class=\\"drac-text drac-text-purple-cyan\\"
      >Outline</span
    ></span
  ><span
    class=\\"drac-badge drac-bg-purple-transparent drac-badge-subtle drac-text-purple\\"
    themecolor=\\"purple\\"
    variant=\\"subtle\\"
    ><span color=\\"purple\\" class=\\"drac-text drac-text-purple\\">Subtle</span></span
  >
</div>
",
  "react": "<div>
  <Badge themeColor=\\"purpleCyan\\" variant=\\"outline\\">
    Outline
  </Badge>
  <Badge themeColor=\\"purple\\" variant=\\"subtle\\">
    Subtle
  </Badge>
</div>;
",
  "title": "Variants",
}
`;
